### ISSUE TITLE
Improve and rename `nla_strlcpy()`

### ISSUE THREAD

**Issue by kees** [KSPP/linux#110]
As part of issue #89, `nla_strlcpy()` could use improvement:

```
/**
 * nla_strlcpy - Copy string attribute payload into a sized buffer
 * @dst: where to copy the string to
 * @nla: attribute to copy the string from
 * @dstsize: size of destination buffer
 *
 * Copies at most dstsize - 1 bytes into the destination buffer.
 * The result is always a valid NUL-terminated string. Unlike
 * strlcpy the destination buffer is always padded out.
 *
 * Returns the length of the source buffer.
 */
```

It seems that `nla_strlcpy()` does tail-padding, like `strscpy_pad()`. It also doesn't do a `strlen()` on the source, instead using `nla_len()`, so this function is at best poorly named. ;) Note also that it can copy _either_ `NLA_STRING` or `NLA_NUL_STRING` (or frankly _anything_ ... is that okay?), and is converting them both into a NUL-terminated string (with tail padding).

The documentation is also wrong: it does not return the length of the source buffer, it returns the length of the source string not including the terminating NUL. If we want this to act like `strscpy_pad()`, it would return what was written (i.e. whether it got truncated: `-E2BIG`). There are exactly 2 callers that actually check the return value, so those should be easy to adjust: both `tcf_proto_check_kind()` and `tcf_action_init1()` are checking for overflow).

It could also use to be made more efficient:

```
        if (dstsize > 0) {
                size_t len = (srclen >= dstsize) ? dstsize - 1 : srclen;

                memset(dst, 0, dstsize);
                memcpy(dst, src, len);
        }
```
There's no reason to 0-fill the first `len` bytes -- they are immediately overwritten.

Coming back to the NLA type (this function will copy _any_ NLA type), but this could lead to confusion for a binary type (for example, if such a binary string included a NUL character: think of the [NUL-byte-in-common-name SSL certs flaw](https://hackaday.com/2009/07/29/black-hat-2009-breaking-ssl-with-null-characters/)).

Most of the callers appear to be working only from `NLA_*STRING`:
```
$ for i in $(git grep -E '\b(nla_strlcpy|tcf_change_indev|rtnl_dev_get|alloc_defdata|tcf_proto_check_kind|nft_obj_lookup|nft_chain_lookup)\b' | grep '\[[A-Z]' | cut -d'[' -f2- | cut -d']' -f1 | sort -u); do git grep '\b'"$i"'\b.*\.type' || echo not found: $i; done
include/net/fib_rules.h:        [FRA_IIFNAME]   = { .type = NLA_STRING, .len = IFNAMSIZ - 1 }, \
include/net/fib_rules.h:        [FRA_OIFNAME]   = { .type = NLA_STRING, .len = IFNAMSIZ - 1 }, \
net/ieee802154/nl_policy.c:     [IEEE802154_ATTR_DEV_NAME] = { .type = NLA_STRING, },
net/decnet/dn_dev.c:    [IFA_LABEL]             = { .type = NLA_STRING,
net/ipv4/devinet.c:     [IFA_LABEL]             = { .type = NLA_STRING, .len = IFNAMSIZ - 1 },
net/core/rtnetlink.c:   [IFLA_ALT_IFNAME]       = { .type = NLA_STRING,
net/core/rtnetlink.c:   [IFLA_IFNAME]           = { .type = NLA_STRING, .len = IFNAMSIZ-1 },
net/core/rtnetlink.c:   [IFLA_INFO_KIND]        = { .type = NLA_STRING },
net/netfilter/ipset/ip_set_hash_netiface.c:             [IPSET_ATTR_IFACE]      = { .type = NLA_NUL_STRING,
net/netfilter/nfnetlink_acct.c: [NFACCT_NAME] = { .type = NLA_NUL_STRING, .len = NFACCT_NAME_MAX-1 },
net/nfc/netlink.c:      [NFC_ATTR_FIRMWARE_NAME] = { .type = NLA_STRING,
net/netfilter/nft_ct.c: [NFTA_CT_HELPER_NAME] = { .type = NLA_STRING,
drivers/infiniband/core/nldev.c:        [RDMA_NLDEV_ATTR_CHARDEV_TYPE]          = { .type = NLA_NUL_STRING,
drivers/infiniband/core/nldev.c:        [RDMA_NLDEV_ATTR_DEV_NAME]              = { .type = NLA_NUL_STRING,
drivers/infiniband/core/nldev.c:        [RDMA_NLDEV_ATTR_LINK_TYPE]             = { .type = NLA_NUL_STRING,
drivers/infiniband/core/nldev.c:        [RDMA_NLDEV_ATTR_NDEV_NAME]             = { .type = NLA_NUL_STRING,
net/sched/act_simple.c: [TCA_DEF_DATA]  = { .type = NLA_STRING, .len = SIMP_MAX_DATA },
net/sched/cls_flower.c: [TCA_FLOWER_INDEV]              = { .type = NLA_STRING,
net/sched/cls_fw.c:     [TCA_FW_INDEV]          = { .type = NLA_STRING, .len = IFNAMSIZ },
net/sched/act_ipt.c:    [TCA_IPT_TABLE] = { .type = NLA_STRING, .len = IFNAMSIZ },
net/sched/sch_api.c:    [TCA_KIND]              = { .type = NLA_STRING },
net/sched/cls_u32.c:    [TCA_U32_INDEV]         = { .type = NLA_STRING, .len = IFNAMSIZ },
net/tipc/netlink.c:     [TIPC_NLA_LINK_NAME]            = { .type = NLA_STRING,
```
But two don't: `fib_metrics_match()` and `ip_metrics_convert()` both use a constructed type, `RTAX_CC_ALGO`, and gets manual reconstruction and size validation, but seems like an overall violation of NLA design. I'm not sure how to fix that one...

I would propose changing the name to either `nla_strcpy()` (shorter) or `nla_strscpy_pad()` (more specific).

So, probably the following logical patch steps:

1. fix the `memset`, `memcpy` inefficiency.
2. change the return value (and callers) to match `strscpy()`, and update docs.
3. rename the function and all callers.
4. refactor to handle the `RTAX_CC_ALGO` case.
5. enforce the nla type to be only `NLA_STRING` or `NLA_NUL_STRING`, and update docs.

**Comment by eiffel-fl** [comment]
Hi.


I will try to go up to step 3 included.
I am not sure I correctly understand step 4 and 5 but maybe things will get clearer by looking deeply into the code.


Best regards.

**Comment by eiffel-fl** [comment]
Hi.


I coded up to step 3.
But I have trouble understanding why `RTAX_CC_ALGO` is a problem for `nla_strlcpy`, can you give me more information please?


Best regards.

**Comment by kees** [comment]
> I coded up to step 3.

Great!

> But I have trouble understanding why `RTAX_CC_ALGO` is a problem for `nla_strlcpy`, can you give me more information please?

I think it'd be nice to have `nla_strscpy_pad()` only usable on "known to be strings" NLA attributes. There are risks to doing a blind conversion of an arbitrary datatype into a string (see my note above about NUL-byte injection). This would be possible except for the very strange `RTAX_CC_ALGO` usage, which is a string, but the entire `RTAX*` set of NLA types appear to be different from standard usage. (i.e. they appear to describe a field _name_, not a field _type_, but maybe I don't understand it well enough.)

For now, I think it is fine to skip steps 4 and 5 until 1, 2, and 3 are landed.

Thanks!

**Comment by eiffel-fl** [comment]
Thank you!

I will review it one more time on my side and I think I will submit it to the mailing list by the end of afternoon.

**Comment by kees** [comment]
v3 series: https://lore.kernel.org/linux-hardening/20201020164707.30402-1-laniel_francis@privacyrequired.com

**Comment by kees** [comment]
https://lore.kernel.org/linux-hardening/20201116092247.608b4f0d@kicinski-fedora-pc1c0hjn.dhcp.thefacebook.com

**Comment by kees** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=872f690341948b502c93318f806d821c56772c42

**Comment by eiffel-fl** [comment]
Coding this two patch series was a pleasure, thank you for the help!

**Comment by kees** [comment]
This was released in v5.11. Thanks again!

**Comment by eiffel-fl** [comment]
Hi.

I am glad this was merged!
It was pleasure to work on it and I thank you for all your advices.


Best regards.
